---
title: "R_assignment"
author: "Ping Kang"
date: "October 9, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Install the packages tidyverse and ggplot
```{r}
if (!require("tidyverse")) install.packages("tidyverse")
library(tidyverse)
if (!require("ggplot2")) install.packages("ggplot2")
library(ggplot2)
```

## Sometimes if some functions do not work, it might be better to reinstall those pakages

```{r}
remove.packages("readr", lib="~/R/x86_64-pc-linux-gnu-library/3.5")
install.packages("readr")
remove.packages("tidyverse", lib="~/R/x86_64-pc-linux-gnu-library/3.5")
install.packages("tidyverse")
library("tidyverse")
```

## Import the files fang_et_al_genotypes.txt and snp_position.txt from R_assignment repository on github

```{r}
fang <- read_tsv("https://raw.githubusercontent.com/pkangISU/R_assignment/master/fang_et_al_genotypes.txt")

snp <- read_tsv("https://raw.githubusercontent.com/pkangISU/R_assignment/master/snp_position.txt")
```

## Inspection of the files of fang_et_al_genotypes.txt and snp_position.txt

```{r}
#check the number of columns in fang and snp
ncol(fang)
ncol(snp)

#check the number of rows in fang and snp
nrow(fang)
nrow(snp)

#check the first six rows of fang and snp
head(fang)
head(snp)

#check the last six rows of fang and snp
tail(fang)
tail(snp)

#check both of the row number and column number at the same time
dim(fang)
dim(snp)

#check the structure of both files
str(fang)
str(snp)

#check the names of both files
names(fang)
names(snp)
```

## Process files to make 20 files of maize and 20 files of teosinte

# pull out maize and teosinte files from fang
  
```{r}

maize <- filter(fang, Group %in% c("ZMMIL","ZMMLR","ZMMMR"))
teosinte <- filter(fang, Group %in% c("ZMPBA","ZMPIL","ZMPJA"))
                   
ncol(maize)
nrow(maize)

ncol(teosinte)
nrow(teosinte)
```

# transpose maize and teosinte files by turning columns to rows

```{r}
maize_t <- t(maize)
ncol(maize_t)
nrow(maize_t)

teosinte_t <- t(teosinte)
ncol(teosinte_t)
nrow(teosinte_t)

head(maize_t)
head(teosinte_t)
```

# pull out three columns from snp files

```{r}
snp_3 <- select(snp, SNP_ID, Chromosome, Position)
ncol(snp_3)
nrow(snp_3)
View(snp_3)
```

# Inspect the data set maize_t

```{r}
str(maize_t)
attributes(maize_t)
is.list(maize_t)
is.vector(maize_t)
is.data.frame(maize_t)
```

# Inspect the data set teosinte_t

```{r}
str(teosinte_t)
attributes(teosinte_t)
is.list(teosinte_t)
is.vector(teosinte_t)
is.data.frame(teosinte_t)
```

# make sure that we have enough memory to run function merge

```{r}
Sys.setenv('R_MAX_VSIZE'=100000000000)
```

# switch maize_t, teosinte_t and snp_3 to data frame type,so that we can use some function

```{r}
maize_t <- as.data.frame(maize_t)
snp_3 <- as.data.frame(snp_3)
teosinte_t <- as.data.frame(teosinte_t)
```

## view the maize_t,snp_3 and teosinte_t files in table to see if they are ready to be merged

```{r}
View(maize_t)
View(snp_3)
View(teosinte_t)
```

## rename column 0 of maize_t as SNP_ID

```{r}
maize_t <- cbind(SNP_ID =rownames(maize_t),maize_t)
colnames(maize_t) <- c("SNP_ID",1:1573)
```

## join maize and snp

```{r}
maize_all <- merge.data.frame(snp_3,maize_t,by="SNP_ID")
ncol(maize_all)
nrow(maize_all)
head(maize_all)
```

# pull out each chromosome
```{r}
maize_chr1 <- filter(maize_all,Chromosome == "1")
nrow(maize_chr1)
maize_chr2 <- filter(maize_all,Chromosome == "2")
maize_chr3 <- filter(maize_all,Chromosome == "3")
maize_chr4 <- filter(maize_all,Chromosome == "4")
maize_chr5 <- filter(maize_all,Chromosome == "5")
maize_chr6 <- filter(maize_all,Chromosome == "6")
maize_chr7 <- filter(maize_all,Chromosome == "7")
maize_chr8 <- filter(maize_all,Chromosome == "8")
maize_chr9 <- filter(maize_all,Chromosome == "9")
maize_chr10 <- filter(maize_all,Chromosome == "10")
```
# sort each chromosome file from maize by increasing position with missing value as ?
```{r}
library(dplyr)
maize_chr1 <- arrange(maize_chr1, as.numeric(Position))
maize_chr2 <-arrange(maize_chr2, as.numeric(Position))
maize_chr3 <-arrange(maize_chr3, as.numeric(Position))
maize_chr4 <-arrange(maize_chr4, as.numeric(Position))
maize_chr5 <-arrange(maize_chr5, as.numeric(Position))
maize_chr6 <-arrange(maize_chr6, as.numeric(Position))
maize_chr7 <-arrange(maize_chr7, as.numeric(Position))
maize_chr8 <-arrange(maize_chr8, as.numeric(Position))
maize_chr9 <-arrange(maize_chr9, as.numeric(Position))
maize_chr10 <-arrange(maize_chr10, as.numeric(Position))
View(maize_chr1)
```
# sort each chromosome file from maize in decreasing position,D stands for decreasing
```{r}
maize_chr1_D <- arrange(maize_chr1,desc(as.numeric(Position)))
View(maize_chr1_D) 
maize_chr2_D <- arrange(maize_chr2,desc(as.numeric(Position)))
maize_chr3_D <- arrange(maize_chr3,desc(as.numeric(Position)))
maize_chr4_D <- arrange(maize_chr4,desc(as.numeric(Position)))
maize_chr5_D <- arrange(maize_chr5,desc(as.numeric(Position)))
maize_chr6_D <- arrange(maize_chr6,desc(as.numeric(Position)))
maize_chr7_D <- arrange(maize_chr7,desc(as.numeric(Position)))
maize_chr8_D <- arrange(maize_chr8,desc(as.numeric(Position)))
maize_chr9_D <- arrange(maize_chr9,desc(as.numeric(Position)))
maize_chr10_D <- arrange(maize_chr10,desc(as.numeric(Position)))
```

# replace missing value ? with - in maize 

```{r}
library(dplyr)
library(stringr)
maize_chr1_Q <- str_replace_all(as.matrix(maize_chr1_D),"\\?","-")
maize_chr1_Q
maize_chr2_Q <- str_replace_all(as.matrix(maize_chr2_D),"\\?","-")
maize_chr3_Q <- str_replace_all(as.matrix(maize_chr3_D),"\\?","-")
maize_chr4_Q <- str_replace_all(as.matrix(maize_chr4_D),"\\?","-")
maize_chr5_Q <- str_replace_all(as.matrix(maize_chr5_D),"\\?","-")
maize_chr6_Q <- str_replace_all(as.matrix(maize_chr6_D),"\\?","-")
maize_chr7_Q <- str_replace_all(as.matrix(maize_chr7_D),"\\?","-")
maize_chr8_Q <- str_replace_all(as.matrix(maize_chr8_D),"\\?","-")
maize_chr9_Q <- str_replace_all(as.matrix(maize_chr9_D),"\\?","-")
maize_chr10_Q <- str_replace_all(as.matrix(maize_chr10_D),"\\?","-")
```

# join teosinte and snp

```{r}
View(teosinte_t)
teosinte_t <- as.data.frame(teosinte_t)
colnames(teosinte_t) <- c (1:975)
teosinte_t <- cbind(SNP_ID =rownames(teosinte_t),teosinte_t)

teosinte_all <- merge (snp_3,teosinte_t,by = "SNP_ID")
ncol(teosinte_all)
nrow(teosinte_all)
View(teosinte_all)
```

# pull out each chromosome from teosinte

```{r}
teosinte_chr1 <- filter(teosinte_all,Chromosome == "1")
nrow(teosinte_chr1)
teosinte_chr2 <- filter(teosinte_all,Chromosome == "2")
teosinte_chr3 <- filter(teosinte_all,Chromosome == "3")
teosinte_chr4 <- filter(teosinte_all,Chromosome == "4")
teosinte_chr5 <- filter(teosinte_all,Chromosome == "5")
teosinte_chr6 <- filter(teosinte_all,Chromosome == "6")
teosinte_chr7 <- filter(teosinte_all,Chromosome == "7")
teosinte_chr8 <- filter(teosinte_all,Chromosome == "8")
teosinte_chr9 <- filter(teosinte_all,Chromosome == "9")
teosinte_chr10 <- filter(teosinte_all,Chromosome == "10")
```

# sort each chromosome file from teosinte by increasing position with missing value as ?

```{r}
library(dplyr)
teosinte_chr1 <- arrange(teosinte_chr1, as.numeric(Position))
teosinte_chr2 <-arrange(teosinte_chr2, as.numeric(Position))
teosinte_chr3 <-arrange(teosinte_chr3, as.numeric(Position))
teosinte_chr4 <-arrange(teosinte_chr4, as.numeric(Position))
teosinte_chr5 <-arrange(teosinte_chr5, as.numeric(Position))
teosinte_chr6 <-arrange(teosinte_chr6, as.numeric(Position))
teosinte_chr7 <-arrange(teosinte_chr7, as.numeric(Position))
teosinte_chr8 <-arrange(teosinte_chr8, as.numeric(Position))
teosinte_chr9 <-arrange(teosinte_chr9, as.numeric(Position))
teosinte_chr10 <-arrange(teosinte_chr10, as.numeric(Position))
View(teosinte_chr1)
```

# sort each chromosome file from teosinte in decreasing position

```{r}
teosinte_chr1_D <- arrange(teosinte_chr1,desc(as.numeric(Position)))
View(teosinte_chr1_D) 
teosinte_chr2_D <- arrange(teosinte_chr2,desc(as.numeric(Position)))
teosinte_chr3_D <- arrange(teosinte_chr3,desc(as.numeric(Position)))
teosinte_chr4_D <- arrange(teosinte_chr4,desc(as.numeric(Position)))
teosinte_chr5_D <- arrange(teosinte_chr5,desc(as.numeric(Position)))
teosinte_chr6_D <- arrange(teosinte_chr6,desc(as.numeric(Position)))
teosinte_chr7_D <- arrange(teosinte_chr7,desc(as.numeric(Position)))
teosinte_chr8_D <- arrange(teosinte_chr8,desc(as.numeric(Position)))
teosinte_chr9_D <- arrange(teosinte_chr9,desc(as.numeric(Position)))
teosinte_chr10_D <- arrange(teosinte_chr10,desc(as.numeric(Position)))
```

# replace missing value ? with - in maize 

```{r}
library(dplyr)
library(stringr)
teosinte_chr1_Q <- str_replace_all(as.matrix(teosinte_chr1_D),"\\?","-")
teosinte_chr1_Q
teosinte_chr2_Q <- str_replace_all(as.matrix(teosinte_chr2_D),"\\?","-")
teosinte_chr3_Q <- str_replace_all(as.matrix(teosinte_chr3_D),"\\?","-")
teosinte_chr4_Q <- str_replace_all(as.matrix(teosinte_chr4_D),"\\?","-")
teosinte_chr5_Q <- str_replace_all(as.matrix(teosinte_chr5_D),"\\?","-")
teosinte_chr6_Q <- str_replace_all(as.matrix(teosinte_chr6_D),"\\?","-")
teosinte_chr7_Q <- str_replace_all(as.matrix(teosinte_chr7_D),"\\?","-")
teosinte_chr8_Q <- str_replace_all(as.matrix(teosinte_chr8_D),"\\?","-")
teosinte_chr9_Q <- str_replace_all(as.matrix(teosinte_chr9_D),"\\?","-")
teosinte_chr10_Q <- str_replace_all(as.matrix(teosinte_chr10_D),"\\?","-")
```

# SNPs per chromosome

```{r}
#maize file
library(ggplot2)
maize_g <- maize_all %>% group_by(Chromosome) %>% summarise(total=n())
View(maize_g)
ggplot(data=maize_g,mapping=aes(x=Chromosome, y=total,fill = Chromosome)) + geom_bar(width = 1,stat = "identity",position = "stack",show.legend = FALSE) + ggtitle("maize snp distribution across chromosomes")

#teosinte file
teosinte_g <- teosinte_all %>% group_by(Chromosome) %>% summarise(total=n())
View(teosinte_g)
ggplot(data=teosinte_g,mapping=aes(x=Chromosome, y=total,fill = Chromosome)) + geom_bar(width = 1,stat = "identity",position = "stack",show.legend = FALSE) + ggtitle("teosinte snp distribution across chromosomes")

```
#what group contributes more?
```{r}
#fang file
library(ggplot2)
View(fang)
fang_all <- fang %>% group_by(Group) %>% summarise(total=n())
View(fang_all)
ggplot(data=fang_all,mapping=aes(x=Group, y=total)) + geom_bar(width = 1,stat = "identity",position = "dodge",show.legend = FALSE)+ facet_wrap( ~ Group)+
  theme_classic() + 
  scale_y_continuous(limit=c(0,1300), expand=c(0,0)) +
  theme(
      axis.title.y=element_text(vjust=0.19),
      axis.title.x=element_text(vjust=0.19),
      legend.position = "bottom",
      legend.title=element_blank()) 

```

# Missing data and amount of heterzygosity and homozygosity

```{r}
#Do not run this part. It is just a record for myself!
#create a new data fang_part including Sample_ID, Group and abph1.20
library(dplyr)
library(stringr)
fang_part <- select(fang, Sample_ID, Group,abph1.20)
nrow(fang_part)
ncol(fang_part)
View(fang_part)
names(fang_part)

```

```{r}
#Do not run this part. I would like to keep this as a record.
#create a new column to show homozygotes and heterozygotes with missing value NA
library(dplyr)
library(stringr)

install.packages("reshape")
library(reshape)

fang_part$type <- TRUE
fang_part$type[fang_part$abph1.20 == "?/?"] <- NA
fang_part$type[fang_part$abph1.20 == "C/C"] <- "Hom"
fang_part$type[fang_part$abph1.20 == "G/G"] <- "Hom"
fang_part$type[fang_part$abph1.20 == "C/G"] <- "Het"
fang_part$type[fang_part$abph1.20 == "G/C"] <- "Het"

names(fang_part)
fang_type <- filter(fang_part, type=="Hom"|type=="Het")

View(fang_type)
unique(fang_type)
distinct(fang_type,Group)
distinct(fang_type,type)

nrow(fang_part)
nrow(fang_type)

```
```{r}
# try to use melt function to generate a image to see the heterozygote, homozygotes and missing value distribution across samples and groups
library(reshape)
library(plyr)
library(dplyr)
library(ggplot2)   #load the packages needed
fang_2E <- fang[,-2]  #exclude column 2 to save space
fang_2E <- as.data.frame(fang_2E) # switch to data.frame
fang_2E[fang_2E[] == "?/?"] <- NA #replace missing values with NA
fang_melt <- melt(fang_2E, id =c("Sample_ID","Group")) #keep 1st and 2nd column, at same time combine the rest of columns by SNP_IDs
fang_melt$type[fang_melt$value == "NA"] <- "NA" #add a new column type with missing value NA 
fang_melt$type[fang_melt$value %in% c("A/A","T/T","C/C","G/G")] <- "TRUE" # assign homozygotes as TRUE
fang_melt$type[fang_melt$value %in% c("A/T","T/A","C/G","G/C","C/T","A/G")] <- "FALSE" # assign heterozygotes as FALSE
fang_melt <- arrange(fang_melt,Group,Sample_ID) #sort the file by Group and Sample_ID

#samples part
fang_sample <- ddply(fang_melt,c("Sample_ID"),summarise,HOM_sample = sum(as.logical(type),na.rm = TRUE),Het_Sample=sum(!as.logical(type), na.rm = TRUE), NA_sample = sum(is.na(type))) #count the number of homozygotes,heterozygotes and missing value by samples
fang_sample_melt <- melt(fang_sample, measure.vars = c("HOM_sample","Het_Sample","NA_sample")) #keep Sample_ID, combine all the other columns

ggplot(fang_sample_melt,aes(x=Sample_ID,y=value,fill=variable))+geom_bar(stat = "identity",position = "fill",size = 1)

# group part
fang_sample <- ddply(fang_melt,c("Group"),summarise,HOM_sample = sum(as.logical(type),na.rm = TRUE),Het_Sample=sum(!as.logical(type), na.rm = TRUE), NA_sample = sum(is.na(type)))
fang_sample_melt <- melt(fang_sample, measure.vars = c("HOM_sample","Het_Sample","NA_sample"))
ggplot(fang_sample_melt,aes(x=Group,y=value,fill=variable))+geom_bar(stat = "identity",position = "fill",size = 1)
                

```

```{r}
#You do not have to look at this part, it is the point plot I was playing with:)
#plot the proportion of homozygous and heterozygous across the groups as well as missing value(2nd graph) or excluding missing values(1st graph)

ggplot(data = fang_type) + geom_point(mapping = aes(x=Group, y=Sample_ID, color=type,stat = "identity",position = "identity",na.rm = FALSE, show.legend = NA,
  inherit.aes = TRUE))
ggplot(data = fang_part) + geom_point(mapping = aes(x=Group, y=Sample_ID, color=type))
```
```{r}
#Do not run this part, it is a record for myself!
#use geom_bar to address the homzygosity and heterozygosity distribution across the groups,only used the SNP location abph1.20
fang_g <- fang_part %>% group_by(Group) %>% count(Sample_ID)
View(fang_g)
fang_m <- merge(fang_part,fang_g,by = "Sample_ID")
View(fang_m)
ggplot(data=fang_m,mapping=aes(x=Group.x, y=n,color=type)) + geom_bar(width = 0.5,stat = "identity",position = "stack",show.legend = TRUE)
```

# My own visualization : see snp distribution across the samples(2nd) and homozygosity distribution on certain SNP location(1st)

```{r}
ggplot(data = fang_part) + geom_bar(mapping = aes(x = type))

ggplot(fang,mapping = aes(x = Sample_ID,y = abph1.20))+geom_point(colour ="blue",size = 3)

```



## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
